File: apps/mobile/modules/sia-os-thumb/android/src/main/java/expo/modules/siaosthumb/SiaOsThumbModule.kt
package expo.modules.siaosthumb

import android.content.ContentUris
import android.graphics.Bitmap
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import android.util.Size
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream
import java.util.UUID

// Writes a system-cached MediaStore thumbnail to the app cache directory
// and resolves with the resulting file:// path. ContentResolver.loadThumbnail
// (API 29+) reads from MediaProvider's on-disk thumbnail cache; the system
// process handles decode and resize. The JPEG encode and write run on the
// AsyncFunction worker thread — no bytes cross the JS bridge.
//
// Returns null on API < 29 and on any failure — callers fall back to the
// in-process resize path.
class SiaOsThumbModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("SiaOsThumb")

    AsyncFunction("getOsThumbnail") { localId: String, targetSize: Double ->
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return@AsyncFunction null
      val context = appContext.reactContext ?: return@AsyncFunction null
      val uri = parseLocalId(localId) ?: return@AsyncFunction null

      val bitmap: Bitmap = try {
        context.contentResolver.loadThumbnail(
          uri,
          Size(targetSize.toInt(), targetSize.toInt()),
          null,
        )
      } catch (_: Exception) {
        return@AsyncFunction null
      }

      val cacheDir = File(context.cacheDir, "sia-os-thumb").apply { mkdirs() }
      val file = File(cacheDir, "${uri.lastPathSegment}-${targetSize.toInt()}-${UUID.randomUUID()}.jpg")
      try {
        FileOutputStream(file).use { bitmap.compress(Bitmap.CompressFormat.JPEG, 85, it) }
      } catch (_: Exception) {
        return@AsyncFunction null
      }

      mapOf(
        "uri" to Uri.fromFile(file).toString(),
        "width" to bitmap.width,
        "height" to bitmap.height,
        "mimeType" to "image/jpeg",
      )
    }
  }

  private fun parseLocalId(localId: String): Uri? {
    if (localId.startsWith("content://")) return Uri.parse(localId)
    val id = localId.toLongOrNull() ?: return null
    return ContentUris.withAppendedId(MediaStore.Files.getContentUri("external"), id)
  }
}

File: apps/mobile/modules/sia-os-thumb/ios/SiaOsThumbModule.swift
import ExpoModulesCore
import Photos
import UIKit

// Writes a system-cached PHAsset thumbnail to the app cache directory and
// resolves with the resulting file:// path. PHImageManager dispatches the
// work to photolibraryd; with deliveryMode=.fastFormat, resizeMode=.exact,
// and isNetworkAccessAllowed=false the call returns a cached tile sized to
// the request without decoding the full asset and without iCloud network.
// The JPEG encode and write also happen on a global queue — no bytes
// touch the JS thread.
//
// Returns nil for any expected failure (asset deleted, permission revoked,
// cached tile unavailable) — callers fall back to the in-process resize.
public class SiaOsThumbModule: Module {
  public func definition() -> ModuleDefinition {
    Name("SiaOsThumb")

    AsyncFunction("getOsThumbnail") { (localId: String, targetSize: Double, promise: Promise) in
      let identifier = localId.hasPrefix("ph://") ? String(localId.dropFirst(5)) : localId
      let fetch = PHAsset.fetchAssets(withLocalIdentifiers: [identifier], options: nil)
      guard let asset = fetch.firstObject else {
        promise.resolve(nil)
        return
      }

      let opts = PHImageRequestOptions()
      opts.deliveryMode = .fastFormat
      opts.resizeMode = .exact
      opts.isNetworkAccessAllowed = false
      opts.isSynchronous = false

      let pixelSize = CGSize(width: targetSize, height: targetSize)

      PHImageManager.default().requestImage(
        for: asset,
        targetSize: pixelSize,
        contentMode: .aspectFill,
        options: opts
      ) { image, _ in
        DispatchQueue.global(qos: .userInitiated).async {
          guard let image = image,
                let data = image.jpegData(compressionQuality: 0.85)
          else {
            promise.resolve(nil)
            return
          }
          let cache = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("sia-os-thumb", isDirectory: true)
          do {
            try FileManager.default.createDirectory(at: cache, withIntermediateDirectories: true)
          } catch {
            promise.resolve(nil)
            return
          }
          let file = cache.appendingPathComponent("\(identifier)-\(Int(targetSize))-\(UUID().uuidString).jpg")
          do {
            try data.write(to: file, options: .atomic)
          } catch {
            promise.resolve(nil)
            return
          }
          promise.resolve([
            "uri": file.absoluteString,
            "width": Int(image.size.width),
            "height": Int(image.size.height),
            "mimeType": "image/jpeg",
          ])
        }
      }
    }
  }
}
